# VCFscope

VCFscope is a tool to investigate and report on the performance of a genome analysis pipeline. It does this by comparing variant calls (in vcf format, with a matched BAM file) generated by an analytical pipeline, to variant calls from a gold standard. Currently this is limited to germline sequencing of NA12878, compared to the GiaB consortium gold-standard reference. This DNAnexus app produces a raw data file (R's RDS format) that is then processed by VCFScope Report to make a detailed human-readable report.

## Inputs
* [Required] vcfgz: A `.vcf.gz` of variants called in an NA12878 sample.
* [Required] bam: A `.bam` of mapped reads used to generate the variants above.
* [Required] bai: A `.bai` index for the BAM file above.
* [Optional] region: A `.bed` file of genomic intervals; only variants overlapping at least one of these intervals will be considered in calculations.  If missing, the entire genome and reportable range are considered.

## Output
* RDS of performance metrics.
