%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%  
%  KCCG WGS Performance Reporter -- Report generator
%  
%  Usage: 
%    Rscript --vanilla -e "library(knitr); knit('report.Rnw', output = 'report.tex')"
%  
%
%  Mark Pinese, 2015
%  
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\documentclass[10pt,a4paper]{article}
\usepackage{geometry}
\usepackage{amsmath}
\usepackage{booktabs}
\usepackage[usenames,dvipsnames,svgnames,table]{xcolor}
\usepackage{url}

\usepackage{fancyhdr}
\setlength{\headheight}{15.2pt}
\pagestyle{fancyplain}
\usepackage{lastpage}

\begin{document}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% PREPARATION
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


% LIBRARIES AND HELPER FUNCTIONS
%--------------------------------------------------------------------
<<load-libs, cache=FALSE, echo=FALSE>>=
suppressMessages(library(VariantAnnotation))
suppressMessages(library(GenomicRanges))
suppressMessages(library(ggplot2))
suppressMessages(library(plyr))
suppressMessages(library(xtable))

suppressMessages(source("report_functions.R"))
@


% LOAD PRECOMPUTED RESULTS
%--------------------------------------------------------------------
<<load-results, cache=FALSE, echo=FALSE>>=
temp = readRDS("report_data.rds")
params = temp$params
class_subsets.performance_thresholded = temp$class_subsets.performance_thresholded
report = temp$report
regions = temp$regions
regions.orig = temp$regions.orig
universe = temp$universe

universe_analysis_size = sum(as.numeric(width(universe$analysis)))
@


% KNITR SETUP
%--------------------------------------------------------------------
<<setup, cache=FALSE, echo=FALSE>>=
library(knitr)
options(
	tikzDocumentDeclaration = "\\documentclass[11pt]{article}",
	tikzLatexPackages = c(
		getOption("tikzLatexPackages"),
		"\\usepackage{amsmath}"),
	tikzMetricsDictionary="tikzMetrics"
)
opts_chunk$set(
	echo = FALSE, results = 'markup', message = FALSE, warning = FALSE, error = TRUE, 
	fig.keep = 'high', fig.align = 'center', fig.width = 5, fig.height = 3.5,
	cache = FALSE, cache.lazy = FALSE)
opts_knit$set(progress = TRUE, verbose = TRUE)
options(warn = 1)
@


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% REPORT STARTS HERE
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\newcommand{\HRule}{\rule{\linewidth}{0.5mm}}

\fancyhf{}
\lhead{WGS Performance Report (v\texttt{\Sexpr{params$version$script}}): \texttt{\Sexpr{texquote(params$sample.id)}}}
\rhead{Page \thepage\ of \pageref{LastPage}}

\let\endtitlepage\relax
\begin{titlepage}
\begin{flushleft}
\LARGE{WGS Performance Report: \texttt{\Sexpr{texquote(params$sample.id)}}}
\HRule
\end{flushleft}
\end{titlepage}

\section{Summary}
Sample ID: \texttt{\Sexpr{texquote(params$sample.id)}} \\
Input VCF: \texttt{\url{\Sexpr{params$path.test.orig}}} \\
Report version: \texttt{\Sexpr{params$version$script}} \\
Report time: \texttt{\Sexpr{texquote(report$gentime)}} \\
Call criterion: \Sexpr{report$criterion_latex}

<<subset-calcs>>=
temp.size_genome = sum(as.numeric(width(universe$genome)))      # | GENOME |
temp.size_subset = sum(as.numeric(width(universe$subset)))      # | BED ^ GENOME |
temp.size_gold = sum(as.numeric(width(universe$analysis)))      # | GIAB ^ BED ^ GENOME |
temp.size_gold_lcr = sum(as.numeric(width(regions$mdust)))      # | GIAB ^ BED ^ GENOME ^ MDUST |

# | (GENOME - GIAB) ^ BED | = | BED - GIAB |
temp.size_notgold = sum(as.numeric(width(setdiff(universe$subset, universe$gold_standard, ignore.strand = TRUE))))

# | (GENOME - GIAB) ^ BED ^ MDUST | = | (BED - GIAB) ^ MDUST |
temp.size_notgold_lcr = sum(as.numeric(width(intersect(setdiff(universe$subset, universe$gold_standard, ignore.strand = TRUE), regions.orig$mdust, ignore.strand = TRUE))))
@

\begin{itemize}
\item Of \Sexpr{temp.size_genome} bases in the genome, \Sexpr{temp.size_subset} (\Sexpr{temp.size_subset / temp.size_genome * 100}\%) were in the target analysis regions.
\item Of \Sexpr{temp.size_subset} bases in the target analysis regions, \Sexpr{temp.size_gold} (\Sexpr{temp.size_gold / temp.size_subset * 100}\%) had gold-standard genotype available.
\item Of \Sexpr{temp.size_gold} target bases with gold-standard calls, \Sexpr{temp.size_gold_lcr} (\Sexpr{temp.size_gold_lcr / temp.size_gold * 100}\%) were in low-complexity regions.
\item Of \Sexpr{temp.size_notgold} target bases without gold-standard calls, \Sexpr{temp.size_notgold_lcr} (\Sexpr{temp.size_notgold_lcr / temp.size_notgold * 100}\%) were in low-complexity regions.
\end{itemize}

<<subset-plots>>=
plotGenomeBreakdown(
    f_targ_of_wg =      temp.size_subset / temp.size_genome, 
    f_gold_of_targ =    temp.size_gold / temp.size_subset,
    f_hc_of_gold =      temp.size_gold_lcr / temp.size_gold,
    f_hc_of_notgold =   temp.size_notgold_lcr / temp.size_notgold
)
@

<<performance-plots>>=
marginalizePerformance = function(perf_data, subset, vars, ...)
{
    ddply(perf_data[eval(subset, perf_data, parent.frame()),], vars, function(x) { 
        ntp = sum(x$ntp)
        nfn = sum(x$nfn)
        nfp = sum(x$nfp)
        ntn = sum(x$ntn)

        if (ntp + nfn + nfp + ntn == 0)
        {
            sens = NA
            sens.lci = NA
            sens.uci = NA
        }
        else
        {
            ci_test = binom.test(ntp, ntp + nfn, ...)
            sens = as.vector(ci_test$estimate)
            sens.lci = ci_test$conf.int[1]
            sens.uci = ci_test$conf.int[2]
        }

        c("sens" = sens, "sens.lci" = sens.lci, "sens.uci" = sens.uci, "n" = ntp + nfn + nfp + ntn)
    }, .drop = TRUE)
}

class_subsets.performance_thresholded$mdust = ordered(c("masked" = "Low Complexity", "unmasked" = "High Complexity")[class_subsets.performance_thresholded$mdust], levels = c("Low Complexity", "High Complexity"))

ggplot(marginalizePerformance(class_subsets.performance_thresholded, quote(muttype == "Subst" & zyg %in% c("RA", "AA")), .(zyg, mdust)), aes(x = zyg, y = sens, fill = zyg)) + 
    geom_bar(stat = "identity", position = "dodge") + 
    geom_errorbar(aes(ymin = sens.lci, ymax = sens.uci), width = 0.2) + 
    labs(x = "", y = "Sensitivity", fill = "Zygosity", title = "SNV Detection Performance") + 
    facet_wrap(~ mdust) + theme_bw() + ylim(0, 1)

ggplot(marginalizePerformance(class_subsets.performance_thresholded, quote(muttype == "Subst" & zyg %in% c("RA", "AA")), .(zyg, mdust)), aes(x = zyg, y = n, fill = zyg)) + 
    geom_bar(stat = "identity", position = "dodge") + 
    labs(x = "", y = "Number of mutations", fill = "Zygosity", title = "SNV Counts in Gold Standard") + 
    facet_wrap(~ mdust) + theme_bw()


ggplot(marginalizePerformance(class_subsets.performance_thresholded, quote(muttype == "Ins" & zyg %in% c("RA", "AA")), .(zyg, mutsize, mdust)), aes(x = mutsize, y = sens, group = zyg, colour = zyg)) + 
    geom_line() + 
    geom_ribbon(aes(ymin = sens.lci, ymax = sens.uci, fill = zyg), alpha = 0.25, colour = NA) + 
    geom_point(size = 2, shape = 21, fill = "white") + 
    labs(x = "Number of bases inserted (inclusive)", y = "Sensitivity", fill = "Zygosity", colour = "Zygosity", title = "Insertion Detection Performance") + 
    facet_wrap(~ mdust, ncol = 1) + theme_bw() + ylim(0, 1)

ggplot(
    marginalizePerformance(class_subsets.performance_thresholded, quote(muttype == "Ins" & zyg %in% c("RA", "AA")), .(zyg, mutsize, mdust)), 
    aes(x = mutsize, y = n, group = zyg, colour = zyg)) + 
    geom_line() + 
    geom_point(size = 2, shape = 21, fill = "white") + 
    labs(x = "Number of bases inserted (inclusive)", y = "Number of mutations", fill = "Zygosity", colour = "Zygosity", title = "Insertion Counts in Gold Standard") + 
    facet_wrap(~ mdust, ncol = 1) + theme_bw()


ggplot(marginalizePerformance(class_subsets.performance_thresholded, quote(muttype == "Del" & zyg %in% c("RA", "AA")), .(zyg, mutsize, mdust)), aes(x = mutsize, y = sens, group = zyg, colour = zyg)) + 
    geom_line() + 
    geom_ribbon(aes(ymin = sens.lci, ymax = sens.uci, fill = zyg), alpha = 0.25, colour = NA) + 
    geom_point(size = 2, shape = 21, fill = "white") + 
    labs(x = "Number of bases deleted (inclusive)", y = "Sensitivity", fill = "Zygosity", colour = "Zygosity", title = "Deletion Detection Performance") + 
    facet_wrap(~ mdust, ncol = 1) + theme_bw() + ylim(0, 1)

ggplot(marginalizePerformance(class_subsets.performance_thresholded, quote(muttype == "Del" & zyg %in% c("RA", "AA")), .(zyg, mutsize, mdust)), aes(x = mutsize, y = n, group = zyg, colour = zyg)) + 
    geom_line() + 
    geom_point(size = 2, shape = 21, fill = "white") + 
    labs(x = "Number of bases deleted (inclusive)", y = "Number of mutations", fill = "Zygosity", colour = "Zygosity", title = "Deletion Counts in Gold Standard") + 
    facet_wrap(~ mdust, ncol = 1) + theme_bw()



false_positive_counts = ddply(subset(class_subsets.performance_thresholded, muttype == "None"), .(mdust), function(x) c(nfp = sum(x$nfp)))
false_positive_counts$subset_size = NA
false_positive_counts$subset_size[false_positive_counts$mdust == "Low Complexity"] = sum(as.numeric(width(regions$mdust)))
false_positive_counts$subset_size[false_positive_counts$mdust == "High Complexity"] = sum(as.numeric(width(setdiff(universe$analysis, regions$mdust, ignore.strand = TRUE))))
stopifnot(sum(false_positive_counts$subset_size) == universe_analysis_size)
false_positive_counts$rate_per_Mb = false_positive_counts$nfp / false_positive_counts$subset_size * 1e6

false_positive_counts$class = NA
false_positive_counts$class[false_positive_counts$mdust == "Low Complexity"] = "Low"
false_positive_counts$class[false_positive_counts$mdust == "High Complexity"] = "High"
false_positive_counts$class = ordered(false_positive_counts$class, levels = c("Low", "High"))
ggplot(false_positive_counts, aes(x = class, y = rate_per_Mb)) + 
    geom_bar(stat = "identity") + 
    labs(x = "Sequence Complexity", y = "False positive rate per Mb", title = "False Positive Rate") + 
    theme_bw()
@

% DATA AND SOFTWARE VERSIONS
%--------------------------------------------------------------------
<<region-version, echo=FALSE, cache=FALSE>>=
temp.region_md5 = "NA"
temp.region_label = "NO"
if (params$region.subset)
{
	temp.region_label = "YES"
	temp.region_md5 = fileMD5(params$region.subset.path)
}
@

\section{Versions}
\begin{itemize}
\item Software: \begin{itemize}
	\item Performance script: \texttt{\Sexpr{texquote(params$version$script)}}
	\item R: \texttt{\Sexpr{texquote(R.version$version.string)}} (\texttt{\Sexpr{texquote(R.version$platform)}})
	\item Genome: \texttt{\Sexpr{texquote(params$genome)}} (\texttt{\Sexpr{packageVersion(params$genome)}})
	\item Java: \texttt{\Sexpr{texquote(params$version$java)}}
	\item RTG core: \texttt{\Sexpr{texquote(params$version$rtg)}}
	\item Bedtools: \texttt{\Sexpr{texquote(params$version$bedtools)}}
	\item Execution host: \texttt{\Sexpr{texquote(params$version$host)}}
	\item Execution time: \Sexpr{texquote(report$gentime)}
\end{itemize}
\item Data: \begin{itemize}
	\item Input VCF: \texttt{\url{\Sexpr{params$path.test.orig}}} (MD5 \texttt{\Sexpr{texquote(fileMD5(params$path.test.orig))}})
	\item Sample ID: \texttt{\Sexpr{texquote(params$sample.id)}}
	\item GiaB VCF: \texttt{\url{\Sexpr{params$path.gold.variants.orig}}} (MD5 \texttt{\Sexpr{texquote(fileMD5(params$path.gold.variants.orig))}})
	\item GiaB BED: \texttt{\url{\Sexpr{params$path.gold.regions.orig}}} (MD5 \texttt{\Sexpr{texquote(fileMD5(params$path.gold.regions.orig))}})
	\item Analysis restricted to regions? \Sexpr{temp.region_label} \begin{itemize}
		\item Analysis region BED: \texttt{\url{\Sexpr{params$region.subset.path}}} (MD5 \texttt{\Sexpr{texquote(temp.region_md5)}})
	\end{itemize}
	\item Sequence masks: \texttt{\url{\Sexpr{params$path.mask.regions.prefix}}} \begin{itemize}
		\item Low complexity: \texttt{\Sexpr{gsub(".*/", "", params$path.mdust)}} (MD5 \texttt{\Sexpr{texquote(fileMD5(params$path.mdust))}})
	\end{itemize}
\end{itemize}

\end{document}
